File: app/src/main/java/com/pr0gramm/app/services/AdminService.java
package com.pr0gramm.app.services;

import com.google.common.collect.ImmutableList;
import com.google.common.primitives.Longs;
import com.pr0gramm.app.api.pr0gramm.Api;
import com.pr0gramm.app.feed.FeedItem;

import javax.annotation.Nullable;
import javax.inject.Inject;
import javax.inject.Singleton;

import gnu.trove.set.TLongSet;
import rx.Observable;

/**
 */
@Singleton
public class AdminService {
    private final Api api;

    @Inject
    public AdminService(Api api) {
        this.api = api;
    }

    /**
     * Deletes an item. This method will not block the user after deleting the item.
     */
    public Observable<Void> deleteItem(FeedItem item, String reason, boolean notifyUser) {
        return deleteItem(item, reason, notifyUser, null);
    }

    public Observable<Api.TagDetails> tagsDetails(long itemId) {
        return api.tagDetails(itemId);
    }

    public Observable<Void> deleteItem(FeedItem item, String reason, boolean notifyUser, @Nullable Float blockDays) {
        String pNotifyUser = notifyUser ? "on" : null;
        String blockUser = blockDays != null && blockDays >= 0 ? "on" : null;
        return api
                .deleteItem(null, item.id(), "custom", reason, pNotifyUser, blockUser, blockDays)
                .map(response -> (Void) null);
    }

    public Observable<Void> deleteTags(long itemId, TLongSet tagIds, @Nullable Float blockDays) {
        String pBlockUser = blockDays != null ? "on" : null;
        return api
                .deleteTag(null, itemId, pBlockUser, blockDays, Longs.asList(tagIds.toArray()))
                .map(response -> (Void) null);
    }

    public static final ImmutableList<String> REASONS = ImmutableList.of(
            "Repost",
            "Auf Anfrage",
            "Regel #1 - Bild unzureichend getagged (nsfw/nsfl)",
            "Regel #1 - Falsche/Sinnlose Nutzung des NSFP Filters",
            "Regel #2 - Gore/Porn/Suggestive Bilder mit Minderjährigen",
            "Regel #3 - Tierporn",
            "Regel #4 - Stumpfer Rassismus/Nazi-Nostalgie",
            "Regel #5 - Werbung/Spam",
            "Regel #6 - Infos zu Privatpersonen",
            "Regel #7 - Bildqualität",
            "Regel #8 - Ähnliche Bilder in Reihe",
            "Regel #12 - Warez/Logins zu Pay Sites",
            "Regel #14 - Screamer/Sound-getrolle",
            "Regel #15 - reiner Musikupload",
            "Trollscheiße.");

}

File: app/src/main/java/com/pr0gramm/app/services/AdminService.kt
package com.pr0gramm.app.services

import com.pr0gramm.app.Logger
import com.pr0gramm.app.api.pr0gramm.Api
import com.pr0gramm.app.feed.FeedItem


/**
 */

class AdminService(private val api: Api, private val cacheService: InMemoryCacheService) {
    private val logger = Logger("AdminService")

    suspend fun tagsDetails(itemId: Long): Api.TagDetails {
        return api.tagDetailsAsync(itemId)
    }

    suspend fun deleteItem(item: FeedItem, reason: String, blockDays: Float? = null) {
        val blockUser = if (blockDays != null && blockDays >= 0) "on" else null
        api.deleteItemAsync(null, item.id, "custom", reason, blockUser, blockDays)
    }

    suspend fun deleteTags(itemId: Long, tagIds: Collection<Long>, blockDays: Float?) {
        cacheService.invalidate()

        val tags = tagIds.toList()

        val pBlockUser = if (blockDays != null) "on" else null
        api.deleteTagAsync(null, itemId, pBlockUser, blockDays, tags)
    }

    suspend fun banUser(name: String, reason: String, days: Float, mode: Api.BanMode) {
        logger.info { "Ban user $name for $days days (mode: $mode): $reason" }
        cacheService.invalidate()

        api.userBanAsync(name, "custom", reason, days, mode)
    }

    suspend fun deleteComment(hard: Boolean, id: Long, reason: String) {
        return if (hard) {
            logger.info { "Doing hard delete of comment $id: $reason" }
            api.hardDeleteCommentAsync(null, id, reason)
        } else {
            logger.info { "Doing soft delete of comment $id: $reason" }
            api.softDeleteCommentAsync(null, id, reason)
        }
    }
}